feat: React navigation app bar guide (#2248)
callstack/react-native-paper: docs/pages/9.react-navigation.md

    </Appbar.Header>
  );
}
```

We want to see application title on main screen and back arrow button on other screens. We can pass navigation props to our `CustomNavigationBar` component. That way we can know if we are on main screen or not.

First let's pass navigation props to `CustomNavigationBar`:

```js
export default function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator
        initialRouteName="Home"
        screenOptions={{
          header: (props) => <CustomNavigationBar {...props} />,
        }}>
        <Stack.Screen name="Home" component={HomeScreen} />
        <Stack.Screen name="Details" component={DetailsScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}
```

Now we can check if navigation have previous state. If it has we can render back arrow button, otherwise we will render app name as before.

```js
function CustomNavigationBar({ navigation, previous }) {
  return (
    <Appbar.Header>
      {previous ? <Appbar.BackAction onPress={navigation.goBack} /> : null}
      <Appbar.Content title="My awesome app" />
    </Appbar.Header>
  );
}
```

![navigationAppBar3](screenshots/react-navigation-appBar3.gif)

> We can easily add additional actions with `Appbar.Action` to our app. For example menu button and let's make it render only on the main screen.

Reviewer: Can we somehow rephrase the second sentence? It's not really understandable to me  :D

---
title: Integrate AppBar with react-navigation
---

# Integrate AppBar with react-navigation

## What we need

 - `react-native-paper`
 - `react-navigation`
 

> We assume that you already have installed libraries above and if that is not the case please check out the guides below.

Reviewer: ```suggestion
We assume that you have already installed the mentioned libraries and if that is not the case please check out the guides below.
```

 
 [React Native Paper - Getting Started](https://callstack.github.io/react-native-paper/getting-started.html)
 
 [React Navigation - Getting Started](https://reactnavigation.org/docs/getting-started/)

## Stack Navigator

> We can start using react-navigation by creating a basic navigation stack. Stack navigator gives us possibility to transition between screens in our app and manage navigation's history. In a simple scenario where there is only one stack navigator present in the app it resembles navigation state in a browser.

Reviewer: ```suggestion
We will start with react-navigation by creating a basic navigation stack. Stack navigator gives us a possibility to transition between screens in our app and manage navigation's history. In a simple scenario where there is only one stack navigator present in the app, it resembles a navigation state in a browser.
```

> When user navigates in the app it pushes or pops items to stack and different screens are displayed.

Reviewer: ```suggestion
When the user navigates in the app, it pushes to stack or pops from the stack and different screens are presented.
```

import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

const Stack = createStackNavigator();

export default function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator initialRouteName="Home">
        <Stack.Screen name="Home" component={HomeScreen} />
        <Stack.Screen name="Details" component={DetailsScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}
```

At the moment our navigation stack contains two screens and will render `HomeScreen` or `DetailsScreen` components according to the current navigation state. We have not implemented those components yet, so let's do this now:

```js
import React from 'react';
import {View, Text, Button, StyleSheet} from 'react-native';

function HomeScreen() {
  return (
    <View style={style.container}>
      <Text>Home Screen</Text>
    </View>
  );
}

function DetailsScreen() {
  return (
    <View style={style.container}>
      <Text>Details Screen</Text>
    </View>
  );
}

const style = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
});
```

> Once we have that we can run the app and check how Stack looks like.

Reviewer: ```suggestion
Once we have finished implementing the components, we can run the app and check how Stack looks like.
```

![navigationAppBar1](screenshots/react-navigation-appBar1.png)

> To navigate from `HomeScreen` to `DetailsScreen` we can use navigation object provided by `Stack.Screen` component. Every component rendered by `Stack.Screen` has access to navigation object via props. Let's modify our `HomeScreen` component:

Reviewer: ```suggestion
To navigate from `HomeScreen` to `DetailsScreen` we can use the navigation object provided by `Stack.Screen` component. Every component rendered by `Stack.Screen` has access to navigation object via props. Let's modify our `HomeScreen` component:
```

```js
function HomeScreen({ navigation }) {
  return (
    <View style={style.container}>
      <Text>Home Screen</Text>
      <Button
        title="Go to details"
        onPress={() => navigation.navigate('Details')}
      />
    </View>
  );
}
```

Our result:

![navigationAppBar2](screenshots/react-navigation-appBar2.gif)

> As you can see, we can already navigate between two screens. In the next steps we will show you how to use Paper's `AppBar` instead of the default header.

Reviewer: ```suggestion
As you can see, we can already navigate between two screens. In the next steps, we will show you how to use Paper's `AppBar` instead of the default header.
```

### Adding AppBar

We can customize Stack's header by passing custom component:

```js
export default function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator
        initialRouteName="Home"
        screenOptions={{
          header: CustomNavigationBar,
        }}>
        <Stack.Screen name="Home" component={HomeScreen} />
        <Stack.Screen name="Details" component={DetailsScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}
```

> Now we can implement `CustomNavigationBar` using `AppBar`:

Reviewer: ```suggestion
Now we will implement `CustomNavigationBar` using `AppBar` component:
```

```js
import { Appbar } from 'react-native-paper';

function CustomNavigationBar() {
  return (
    <Appbar.Header>
      <Appbar.Content title="My awesome app" />
    </Appbar.Header>
  );
}
```

> We want to see application title on the main screen and back arrow button on other screens. We can pass navigation props to our `CustomNavigationBar` component. This way we can know if we are on the main screen or not.

Reviewer: ```suggestion
Our current implementation of the `CustomNavigationBar` is very simple, we just render a title inside of it. There is currently no way to go back to the previous screen because the back button is not visible in the header. Let's add it now and let's make sure it's visible on all Stack's screens except `Home` screen.
```

> First let's pass navigation props to `CustomNavigationBar`:

Reviewer: ```suggestion
First, let's pass navigation props to `CustomNavigationBar`:
```

> First, let's pass navigation props to `CustomNavigationBar`:

Reviewer 2: ```suggestion
Firstly, pass navigation props to `CustomNavigationBar`:
```

```js
export default function App() {
  return (
    <NavigationContainer>
      <Stack.Navigator
        initialRouteName="Home"
        screenOptions={{
          header: (props) => <CustomNavigationBar {...props} />,
        }}>
        <Stack.Screen name="Home" component={HomeScreen} />
        <Stack.Screen name="Details" component={DetailsScreen} />
      </Stack.Navigator>
    </NavigationContainer>
  );
}
```

> For the next step we check if navigation has previous state. If it has, we can render back arrow button, otherwise we will render app name as before.

Reviewer: ```suggestion
As a second step, we check if navigation has previous state. If it has, it means there is another screen on the stack beneath the current screen and we should render the back arrow button in such a case.
```

> As a second step, we check if navigation has previous state. If it has, it means there is another screen on the stack beneath the current screen and we should render the back arrow button in such a case.

Reviewer 2: ```suggestion
Secondly, check if navigation has previous state. If it has, it means there is another screen on the stack beneath the current screen and we should render the back arrow button in such a case.
```

```js
function CustomNavigationBar({ navigation, previous }) {
  return (
    <Appbar.Header>
      {previous ? <Appbar.BackAction onPress={navigation.goBack} /> : null}
      <Appbar.Content title="My awesome app" />
    </Appbar.Header>
  );
}
```

![navigationAppBar3](screenshots/react-navigation-appBar3.gif)

Another interesting pattern that can be implemented with React-Native-Paper and React-Navigation is a "hamburger menu". Thanks to the Menu component we can add a nice looking pop-up to our Appbar. To implement this feature we need to make a couple of changes in `CustomNavigationBar`:
- Render a `Menu` component
- Pass `Appbar.Action` to the anchor prop
- Add a state to control `Menu` visibility

> We also want our manu to appear only on `HomeScreen`, so we will check if `previous` props is available:

Reviewer: ```suggestion
We also want the menu to appear only on `HomeScreen`, which means we will render it conditionally based on the `previous` prop.
```

> Another interesting pattern that can be implemented with React-Native-Paper and React-Navigation is a "hamburger menu". Thanks to the Menu component we can add a nice looking pop-up to our Appbar. To implement this feature we need to make a couple of changes in `CustomNavigationBar`:

Reviewer 2: ```suggestion
Another interesting pattern that can be implemented with `react-native-paper` and `react-navigation` is a "hamburger menu". Thanks to the `Menu` component we can add a nice looking pop-up to our `Appbar`. To implement this feature we need to make a couple of changes in `CustomNavigationBar`:
```

![navigationAppBar3](screenshots/react-navigation-appBar3.gif)

Another interesting pattern that can be implemented with React-Native-Paper and React-Navigation is a "hamburger menu". Thanks to the Menu component we can add a nice looking pop-up to our Appbar. To implement this feature we need to make a couple of changes in `CustomNavigationBar`:
- Render a `Menu` component
- Pass `Appbar.Action` to the anchor prop
- Add a state to control `Menu` visibility

> We also want our manu to appear only on `HomeScreen`, so we will check if `previous` props is available:

Reviewer 3: Typo: manu -> menu
